### PR TITLE
gcp/observability: implement public preview config syntax, logging schema, and exposed metrics

### PR DESCRIPTION
**Pull request by zasweq**
This PR implements all the changes defined for public preview. The biggest change is the configuration change, which switched the logging from precedence based to iterative based and added a partition between client and server RPC events. The second one is the logging schema change defined by Eric, which I used an internal go struct to represent. The final trivial change is the change in the metrics collected/views registered to only Completed RPCs.

RELEASE NOTES:
* gcp/observability: support new configuration defined in public preview user guide

### CODE REVIEW

File: gcp/observability/opencensus.go [hunk 1]
```diff
@@ -128,9 +127,6 @@ func startOpenCensus(config *config) error {
 // packages if exporter was created.
 func stopOpenCensus() {
 	if exporter != nil {
-		internal.ClearGlobalDialOptions()
-		internal.ClearGlobalServerOptions()
```
Comment (dfawley) [inline gcp/observability/opencensus.go:132]: Why were these moved?  The global dial/server options are set in `startOpenCensus`, so it seems like we should be clearing them here (even if we need to do it unconditionally).

File: internal/binarylog/binarylog.go [hunk 2]
```diff
@@ -161,11 +161,11 @@ func (l *logger) setBlacklist(method string) error {
 	return nil
 }
 
-// getMethodLogger returns the methodLogger for the given methodName.
+// getMethodLogger returns the BinMethodLogger for the given methodName.
```
Comment (dfawley) [inline internal/binarylog/binarylog.go:164]: This still returns `MethodLogger`.  Why the change to `BinMethodLogger` in the comment?

File: internal/binarylog/method_logger.go [hunk 3]
```diff
@@ -52,7 +52,9 @@ type MethodLogger interface {
 	Log(LogEntryConfig)
 }
 
-type methodLogger struct {
+// BinMethodLogger is a method logger that truncates headers and messages based
+// on configured fields.
+type BinMethodLogger struct {
```
Comment (dfawley) [inline internal/binarylog/method_logger.go:57]: Nit: `TruncatingMethodLogger` instead?

**Review by dfawley — COMMENTED** [review]
Mostly nits so far, but I haven't dug into the tests.

We need to get this running in an integration type of test ASAP to make sure we figure out if it's using the proper JSON field names for the config and log entries being exported.

File: gcp/observability/opencensus.go (hunk 1 above)
Comment (zasweq) [reply gcp/observability/opencensus.go:132]: Fair question. Now, global dial and server options are being added in logging.go. Unlike this file, while holds a ref to a global exporter, which can be used as an invariant to determine whether global dial options and server options have been registered in this package, logging.go has no such mechanism of determining to add a conditional check on whether they have been registered. Thus, I found it best to simply clear unconditionally in observability.go, encompassing both the cleanup for logging (which we can't do conditionally), and also for openCensus (which we can do conditionally.

File: internal/binarylog/binarylog.go (hunk 2 above)
Comment (zasweq) [reply internal/binarylog/binarylog.go:164]: Well it sort of returns a BinMethodLogger, which implements MethodLogger (the actual return type). Switched back. Also, the original comment was referring to the specific implementation (methodLogger vs. MethodLogger).

File: internal/binarylog/method_logger.go (hunk 3 above)
Comment (zasweq) [reply internal/binarylog/method_logger.go:57]: Sure. Switched.

File: gcp/observability/config.go [hunk 4]
```diff
@@ -169,13 +127,133 @@ func parseObservabilityConfig() (*config, error) {
...
+	//
+	// The service name, when specified, must be the fully qualified service
+	// name, including the package name.
+	//
+	// Examples:
+	//    1."goo.Foo/Bar" selects only the method "Bar" from service "goo.Foo",
+	//    here “goo” is the package name.
+	//    2."goo.Foo/*" selects all methods from service "goo.Foo"
+	//    3. "*" selects all methods from all services.
+	Method []string `json:"method,omitempty"`
```
[the remark this answers is not in the document]
Comment (zasweq) [reply gcp/observability/config.go:201]: Switched. Saw your comment but figured I'd wait until the review pass.

File: gcp/observability/config.go [hunk 5]
```diff
@@ -169,13 +127,133 @@ func parseObservabilityConfig() (*config, error) {
...
```
[the remark this answers is not in the document]
Comment (zasweq) [reply gcp/observability/config.go:163]: Switched. Saw your comment but figured I'd wait until the review pass.

File: gcp/observability/config.go [hunk 6]
```diff
@@ -123,14 +61,34 @@ func fetchDefaultProjectID(ctx context.Context) string {
...
+		match := methodStringRegexp.FindStringSubmatch(method)
 		if match == nil {
-			return fmt.Errorf("invalid log filter Pattern: %v", filter.Pattern)
+			return fmt.Errorf("invalid method string: %v", method)
+		}
+	}
+	return nil
+}
+
+func validateFilters(config *config) error {
```
[the remark this answers is not in the document]
Comment (zasweq) [reply gcp/observability/config.go:80]: Actually, I just realized something. "logFilters" is carried over from Lidi's config, as there is no such concept about filters in the new config. I switched this to "validateLoggingEvents", as the config defines them as client/server RPC events. CloudLogging seems wrong here, this is specific to our layer of logic, which sure will be exported to cloud logging but that name doesn't feel right.

File: gcp/observability/config.go [hunk 7]
```diff
@@ -123,14 +61,34 @@ func fetchDefaultProjectID(ctx context.Context) string {
 	return credentials.ProjectID
 }
 
-func validateFilters(config *config) error {
-	for _, filter := range config.LogFilters {
-		if filter.Pattern == "*" {
+func validateMethods(methods []string, exclude bool) error {
```
[the remark this answers is not in the document]
Comment (zasweq) [reply gcp/observability/config.go:64]: See comment below; switched to validateLogEventMethod. Let me know if these names don't work for you.

File: gcp/observability/config.go [hunk 8]
```diff
@@ -169,13 +127,133 @@ func parseObservabilityConfig() (*config, error) {
...
+	// the size of the metadata is greater than the defined limit, content past
+	// the limit will be truncated. The default value is 0.
+	MaxMetadataBytes int `json:"max_metadata_bytes"`
+	// MaxMessageBytes is the maximum number of bytes of each message to log. If
+	// the size of the message is greater than the defined limit, content past
+	// the limit will be truncated. The default value is 0.
+	MaxMessageBytes int `json:"max_message_bytes"`
+}
+
+type serverRPCEvents struct {
```
[the remark this answers is not in the document]
Comment (zasweq) [reply gcp/observability/config.go:179]: Completely fair point. However, do you see the logic diverging ever between client/server RPC events. If so, I would prefer to keep as such for maintainability. If not, I can make this change. Would like to hear your answer to that question first though :).

**Review by zasweq — COMMENTED** [review]
Thanks for the comments boss man :D!

File: internal/binarylog/binarylog.go (hunk 2 above)
Comment (dfawley) [reply internal/binarylog/binarylog.go]: If it always returns a `BinMethodLogger` then should we change the signature of the function?  Or it's that way to satisfy an interface and we can't change the interface?

File: gcp/observability/go.mod [hunk 9]
```diff
@@ -5,12 +5,11 @@ go 1.14
 require (
 	cloud.google.com/go/logging v1.4.2
 	contrib.go.opencensus.io/exporter/stackdriver v0.13.12
-	github.com/golang/protobuf v1.5.2
+	github.com/google/go-cmp v0.5.6
```
[the remark this answers is not in the document]
Comment (dfawley) [reply gcp/observability/go.mod:8]: OMG my eyes.  I thought you bumped the version from x.5.2 to x.5.6 of go-cmp.  Pretend I didn't say anything here.

File: gcp/observability/config.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (dfawley) [reply gcp/observability/config.go:179]: That's fair.  It's possible the client could get a `authority`/`target` field to filter based on that, which wouldn't make sense on the server.  Pretty unfortunate to have 2 copies of everything until that theoretical "one day" though.

File: gcp/observability/opencensus.go (hunk 1 above)
Comment (dfawley) [reply gcp/observability/opencensus.go:132]: We could always clear it in both places.  It shouldn't be an error to clear it if it's already clear.

Minor point, but it's nice if `stopSomething` undoes `startSomething` instead of relying on adjacent code.

File: internal/binarylog/binarylog.go (hunk 2 above)
Comment (zasweq) [reply internal/binarylog/binarylog.go]: Discussed offline; need this method signature to satisfy interface. I did think some comments were incorrect though (in master) and switched them.

File: gcp/observability/opencensus.go (hunk 1 above)
Comment (zasweq) [reply gcp/observability/opencensus.go]: Discussed offline; left opencensus.go as is and cleared within the exporter != nil invariant, and logging.go cleared server/dial options unconditionally.

File: gcp/observability/config.go [hunk 10]
```diff
@@ -34,73 +35,10 @@ const (
 	envObservabilityConfig     = "GRPC_CONFIG_OBSERVABILITY"
 	envObservabilityConfigJSON = "GRPC_CONFIG_OBSERVABILITY_JSON"
```
Comment (dfawley) [inline gcp/observability/config.go:36]: The only env vars we have now are `GRPC_OBSERVABILITY_CONFIG` and `GRPC_OBSERVABILITY_CONFIG_FILE`.

Can we add these into envconfig instead?

File: gcp/observability/logging_test.go [hunk 11]
```diff
@@ -0,0 +1,877 @@
...
+	gcplogging "cloud.google.com/go/logging"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	binlogpb "google.golang.org/grpc/binarylog/grpc_binarylog_v1"
+	"google.golang.org/grpc/internal/stubserver"
+	"google.golang.org/grpc/test/grpc_testing"
+)
+
+func cmpLoggingEntryList(entryList1 []*grpcLogEntry, entryList2 []*grpcLogEntry) error {
+	if diff := cmp.Diff(entryList1, entryList2,
```
Comment (dfawley) [inline gcp/observability/logging_test.go:40]: Please name these `got` and `want` so it's clear how to call it when adding a new test case.

File: gcp/observability/logging_test.go [hunk 12]
```diff
@@ -0,0 +1,877 @@
...
+	defer cleanup()
+
+	ss := &stubserver.StubServer{
+		UnaryCallF: func(ctx context.Context, in *grpc_testing.SimpleRequest) (*grpc_testing.SimpleResponse, error) {
+			return &grpc_testing.SimpleResponse{}, nil
+		},
+		FullDuplexCallF: func(stream grpc_testing.TestService_FullDuplexCallServer) error {
+			for {
+				_, err := stream.Recv()
+				if err == io.EOF {
```
Comment (dfawley) [inline gcp/observability/logging_test.go:135]: Send a message here to make sure `ServerMessage` works for streaming RPCs?

File: gcp/observability/logging_test.go [hunk 13]
```diff
@@ -0,0 +1,877 @@
...
```
Comment (dfawley) [inline gcp/observability/logging_test.go:310]: As above; send a message?

Also it seems worthwhile to send a message from the client in both cases.

File: gcp/observability/logging_test.go [hunk 14]
```diff
@@ -0,0 +1,877 @@
...
+	// accordingly. The first event it matches to should be used for the
+	// configuration, even though it could potentially match to events in the
+	// future.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if _, err := ss.Client.UnaryCall(ctx, &grpc_testing.SimpleRequest{}); err != nil {
+		t.Fatalf("Unexpected error from UnaryCall: %v", err)
+	}
+	grpcLogEntriesWant := make([]*grpcLogEntry, 5)
+	grpcLogEntriesWant[0] = &grpcLogEntry{
```
Comment (dfawley) [inline gcp/observability/logging_test.go:612]: FWIW you can initialize these like:

```go
grpcLogEntriesWant := []*grpcLogEntry{
	{
		Type: eventTypeClientHeader,
		Logger: ....
		...
	}, {
		Type: eventTypeClientMessage,
		...
	}
}
```

**Review by dfawley — COMMENTED** [review]
Almost fully LGTM.  Mostly just some small things.

Please also add a test that configures payload and metadata truncation and verifies the output events.

File: gcp/observability/config.go [hunk 15]
```diff
@@ -123,14 +61,34 @@ func fetchDefaultProjectID(ctx context.Context) string {
...
-func validateFilters(config *config) error {
-	for _, filter := range config.LogFilters {
-		if filter.Pattern == "*" {
+func validateMethods(methods []string, exclude bool) error {
+	for _, method := range methods {
+		if method == "*" {
+			if exclude {
+				return errors.New("cannot have exclude and a '*' wildcard")
+			}
 			continue
```
[the remark this answers is not in the document]
Comment (dfawley) [reply gcp/observability/config.go:70]: Ping on this one.

[the remark this answers is not in the document]
Comment (zasweq) [reply gcp/observability/config.go:70]: Oh sorry forgot to respond to this earlier (I swear I typed something out at least - it must've gotten lost). No, there was not supposed to be an error here. Previously, yes this was speced out somewhere (was a docstring in the config type by Lidi). However, now we decided not to do any preprocessing (me and you had a discussion about this) and left it up to the user to know if he specifies a * somewhere in method[] for an event, no further event will hit, and entrusted the user with more responsibility in that regard, and trusted the user.

File: gcp/observability/logging_test.go (hunk 13 above)
Comment (zasweq) [reply gcp/observability/logging_test.go:310]: Also, something I just realized is this takes away having to change this test in the future due to that bin logging bug of logging Server Headers on Trailers only response. Luckily we have it clearly documented in other PRs test and this PRs history.

**Review by zasweq — COMMENTED** [review]
Thanks for the pass bossman :D! Got to all your comments and added a headers/message truncated test.

File: gcp/observability/logging_test.go [hunk 16]
```diff
@@ -131,8 +133,13 @@ func (s) TestClientRPCEventsLogAll(t *testing.T) {
 		},
 		FullDuplexCallF: func(stream grpc_testing.TestService_FullDuplexCallServer) error {
 			for {
```
Comment (dfawley) [inline gcp/observability/logging_test.go:135]: I don't think this actually wants to be in a loop.  The client just sends once, right?

File: gcp/observability/logging_test.go [hunk 17]
```diff
@@ -306,8 +341,13 @@ func (s) TestServerRPCEventsLogAll(t *testing.T) {
...
```
Comment (dfawley) [inline gcp/observability/logging_test.go:343]: Similar re: not using a loop here.

File: gcp/observability/logging_test.go [hunk 18]
```diff
@@ -671,7 +671,7 @@ func (s) TestClientRPCEventsTruncateHeaderAndMetadata(t *testing.T) {
 			Authority:   ss.Address,
 			SequenceID:  1,
 			Payload: payload{
-				Metadata: map[string]string{"key1": "value1"},
+				Metadata: map[string]string{},
```
Comment (dfawley) [inline gcp/observability/logging_test.go:674]: This is definitely unfortunate that we don't get anything here.  It would be better if we could accept either header.  There should be a way to do that in the `cmp` package.

File: gcp/observability/config.go [hunk 19]
```diff
@@ -113,14 +112,14 @@ func unmarshalAndVerifyConfig(rawJSON json.RawMessage) (*config, error) {
 }
 
 func parseObservabilityConfig() (*config, error) {
-	if fileSystemPath := os.Getenv(envObservabilityConfigJSON); fileSystemPath != "" {
-		content, err := ioutil.ReadFile(fileSystemPath) // TODO: Switch to os.ReadFile once dropped support for go 1.15
+	if envconfig.ObservabilityConfigFile != "" {
```
[the remark this answers is not in the document]
Comment (zasweq) [reply gcp/observability/config.go:115]: The else if below the "long exported variable" is used once so I think it's fair there.

File: gcp/observability/config.go [hunk 20]
```diff
@@ -113,14 +112,14 @@ func unmarshalAndVerifyConfig(rawJSON json.RawMessage) (*config, error) {
...
+		content, err := ioutil.ReadFile(envconfig.ObservabilityConfigFile) // TODO: Switch to os.ReadFile once dropped support for go 1.15
```
[the remark this answers is not in the document]
Comment (zasweq) [reply gcp/observability/config.go:116]: Done, except with GCP added.

File: gcp/observability/logging_test.go (hunk 16 above)
Comment (zasweq) [reply gcp/observability/logging_test.go:135]: Ah, lol fair point :). Lol unnecessary while true.

File: gcp/observability/logging_test.go (hunk 17 above)
Comment (zasweq) [reply gcp/observability/logging_test.go:343]: Got rid of every while true loop in all my FullDuplexCallF definitions.

File: gcp/observability/logging_test.go (hunk 18 above)
Comment (zasweq) [reply gcp/observability/logging_test.go:674]: Ok, I looked through cmp/cmpopts, and I couldn't find an explicit option that had the behavior you desire. The solution I came up with is to ignore the metadata in the cmp.Diff, and then compare with an or explicitly afterward.

File: gcp/observability/config.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (zasweq) [reply gcp/observability/config.go:116]: Also I'm assuming you wanted this call on the logger, not the package grpclog? Used the logger.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments bossman :D!

**Review by dfawley — APPROVED** [review]
Please check len() of metadata in the test that is expecting truncation.  And fix the other final issue.  Otherwise LGTM!
